### PR TITLE
Fix handling of read enables for write-first (default) memories in VerilogMemDelays

### PR DESCRIPTION
**Pull request by albert-magyar**
Kudos to @chick for exposing a lurking bug in his issue #1179.

By default, `VerilogMemDelays` has always generated write-first memories with the pipeline registers it adds to the memory access commands. The way this was being handled had two issues:

- For normal sync-read mems (`read-latency` and `write-latency` both 1), feeding back read data to an enable field would result in a combinational loop. This came up for the cache of riscv-mini in #1179, but there is a much smaller non-working example that is now fixed in this PR (and included as a test).

- For any memories with `read-latency` greater than 1, the enable signal does not behave correctly, and can lead to incorrect read data. This space was not covered very well by existing Verilog tests.

I added some tests to try to cover some of this space. Several of of these tests (the ones with write-first, two-cycle-read memories) fail without the fix to `VerilogMemDelays`.

Some test infrastructure for getting boilerplate out of these tests ended up in a library exposing a chisel-testers2-like API (poke, expect, step) for simple Verilog execution tests. I tried to write some Scaladoc for the parts that might be reusable, and I put them in the `firrtlTests.execution` package.

Minimal example that caused combinational loops before this PR:
```
circuit loop :
  module loop :
    input clk : Clock
    input raddr : UInt<5>
    mem m :
      data-type => UInt<1>
      depth => 32
      reader => r
      read-latency => 1
      write-latency => 1
      read-under-write => undefined
    m.r.clk <= clk
    m.r.addr <= raddr
    m.r.en <= m.r.data
```

### CODE REVIEW

File: src/test/scala/firrtlTests/MemEnFeedbackSpec.scala [hunk 1]
```diff
@@ -0,0 +1,41 @@
+// See LICENSE for license details.
```
Comment (ekiwi) [inline src/test/scala/firrtlTests/MemEnFeedbackSpec.scala]: Should this go into a "regressions" package?
In general this mostly seem to be testing that the combinatorial loop bug is fixed.
It is not a generic test of sync-read memories.

File: src/test/scala/firrtlTests/MemLatencySpec.scala [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+  def writeOnly(addr: Int, data: Int) = MemAccess(Some(Write(addr, data)), None)
+  def readOnly(addr: Int, expectedValue: Int) = MemAccess(None, Some(Read(addr, expectedValue)))
+}
+
+abstract class MemLatencySpec(rLatency: Int, wLatency: Int, ruw: String)
+    extends SimpleExecutionTest
+    with VerilogExecution {
+
+  import MemLatencySpec._
+
```
Comment (ekiwi) [inline src/test/scala/firrtlTests/MemLatencySpec.scala]: Maybe add:
```
assert(wLatency >= 1)
assert(rLatency >= 0)
```

File: src/test/scala/firrtlTests/MemLatencySpec.scala [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+       |  data-type => UInt<32>
+       |  depth => 256
+       |  reader => r
+       |  writer => w
+       |  read-latency => ${rLatency}
+       |  write-latency => ${wLatency}
+       |  read-under-write => ${ruw}
+       |m.r.clk <= clock
+       |m.w.clk <= clock
+       |m.w.mask <= UInt<1>(1)
```
Comment (ekiwi) [inline src/test/scala/firrtlTests/MemLatencySpec.scala]: Maybe we can also test the write mask:
https://github.com/freechipsproject/treadle/blob/5071f74d6092b30023e55cd56e3f0ebe39736baa/src/test/scala/treadle/MemoryUsageSpec.scala#L472

File: src/test/scala/firrtlTests/MemLatencySpec.scala [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+class WriteFirstMemReadEnTogglesRL2 extends MemLatencySpec(rLatency = 2, wLatency = 1, ruw = "new") with ReadEnToggles
+class ReadFirstMemReadEnTogglesRL2 extends MemLatencySpec(rLatency = 2, wLatency = 1, ruw = "old") with ReadEnToggles
+
+// Write latency 2
+class WriteFirstMemReadEnTogglesWL2 extends MemLatencySpec(rLatency = 1, wLatency = 2, ruw = "new") with ReadEnToggles
+class ReadFirstMemReadEnTogglesWL2 extends MemLatencySpec(rLatency = 1, wLatency = 2, ruw = "old") with ReadEnToggles
+
+// Read latency 2, write latency 2
+class WriteFirstMemReadEnTogglesRL2WL2 extends MemLatencySpec(rLatency = 2, wLatency = 2, ruw = "new") with ReadEnToggles
+class ReadFirstMemReadEnTogglesRL2WL2 extends MemLatencySpec(rLatency = 2, wLatency = 2, ruw = "old") with ReadEnToggles
```
Comment (ekiwi) [inline src/test/scala/firrtlTests/MemLatencySpec.scala]: This kind of test is very cool!
Nice job!

**Comment by ekiwi** [comment]
Seems like yosys is crashing in the CI.

File: src/test/scala/firrtlTests/MemEnFeedbackSpec.scala (hunk 1 above)
Comment (albert-magyar) [reply src/test/scala/firrtlTests/MemEnFeedbackSpec.scala]: I could see arguments both ways (it is testing a corner case), but this type of test is currently interspersed among many of the other test files. Finding and moving them might make sense in a later PR.

**Comment by albert-magyar** [comment]
> Seems like yosys is crashing in the CI.

Yes, it's not actually formally equivalent (and Yosys is running out of memory before it can fail), but that is just exposing that LEC provides an imperfect abstraction for a formal spec. The cache response data for the `ICache` design differs when the cache is *not* active.

I am actually working on some changes to the emitted microarchitecture that may help QoR for some of the less-used cases (read-first, etc.). I will also make the design formally equivalent for the parameterization used in the cache by incorporating extra gating, so this is not quite ready.

File: src/test/scala/firrtlTests/MemLatencySpec.scala (hunk 2 above)
Comment (albert-magyar) [reply src/test/scala/firrtlTests/MemLatencySpec.scala]: That seems like a good idea.

File: src/test/scala/firrtlTests/MemLatencySpec.scala (hunk 3 above)
Comment (albert-magyar) [reply src/test/scala/firrtlTests/MemLatencySpec.scala]: Sure, why not :)

File: src/test/scala/firrtlTests/MemLatencySpec.scala (hunk 4 above)
Comment (albert-magyar) [reply src/test/scala/firrtlTests/MemLatencySpec.scala]: Thanks

File: src/test/scala/firrtlTests/MemLatencySpec.scala (hunk 3 above)
Comment (albert-magyar) [reply src/test/scala/firrtlTests/MemLatencySpec.scala]: I added a mask-0, en-1 write operation that checks that the previous value is maintained. Adding more elaborate cases would require complicating how the test is structured, but that at least adds a nice sanity check for the wide latency space.

File: src/test/scala/firrtlTests/MemLatencySpec.scala (hunk 2 above)
Comment (albert-magyar) [reply src/test/scala/firrtlTests/MemLatencySpec.scala]: I added `require`s

**Comment by albert-magyar** [comment]
CI formal equivalence was choking on the fact that some registers changed names, so I just reverted the naming of memory command / response pipeline registers to what it was before.

**Comment by jackkoenig** [comment]
Is this `1.2.1` or `1.3.0`?

**Comment by albert-magyar** [comment]
This should definitely be backported. I turned on the 1.2.1 milestone, but feel free to change/label whatever on this.

**Comment by jackkoenig** [comment]
This breaks binary compatibility:

```
[error]  * class firrtl.passes.memlib.DelayPipe does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[MissingClassProblem]("firrtl.passes.memlib.DelayPipe")
[error]  * class firrtl.passes.memlib.DelayPipe#PipeState does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[MissingClassProblem]("firrtl.passes.memlib.DelayPipe$PipeState")
[error]  * method ug()firrtl.UnknownFlow# in object firrtl.passes.memlib.VerilogMemDelays does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("firrtl.passes.memlib.VerilogMemDelays.ug")
[error]  * method expToString(firrtl.ir.Expression)java.lang.String in object firrtl.passes.memlib.VerilogMemDelays does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("firrtl.passes.memlib.VerilogMemDelays.expToString")
[error]  * method buildNetlist(scala.collection.mutable.HashMap,firrtl.ir.Statement)Unit in object firrtl.passes.memlib.VerilogMemDelays does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("firrtl.passes.memlib.VerilogMemDelays.buildNetlist")
[error]  * method memDelayStmt(scala.collection.mutable.HashMap,firrtl.Namespace,scala.collection.mutable.HashMap,scala.collection.mutable.ArrayBuffer,firrtl.ir.Statement)firrtl.ir.Statement in object firrtl.passes.memlib.VerilogMemDelays does not have a correspondent in current version                            
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("firrtl.passes.memlib.VerilogMemDelays.memDelayStmt")
[error]  * method replaceExp(scala.collection.mutable.HashMap,firrtl.ir.Expression)firrtl.ir.Expression in object firrtl.passes.memlib.VerilogMemDelays does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("firrtl.passes.memlib.VerilogMemDelays.replaceExp")
[error]  * method appendStmts(scala.collection.Seq,firrtl.ir.Statement)firrtl.ir.Statement in object firrtl.passes.memlib.VerilogMemDelays does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("firrtl.passes.memlib.VerilogMemDelays.appendStmts")
[error]  * method memDelayMod(firrtl.ir.DefModule)firrtl.ir.DefModule in object firrtl.passes.memlib.VerilogMemDelays does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("firrtl.passes.memlib.VerilogMemDelays.memDelayMod")
[error]  * object firrtl.passes.memlib.DelayPipe does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[MissingClassProblem]("firrtl.passes.memlib.DelayPipe$")
[error]  * static method memDelayMod(firrtl.ir.DefModule)firrtl.ir.DefModule in class firrtl.passes.memlib.VerilogMemDelays does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("firrtl.passes.memlib.VerilogMemDelays.memDelayMod")
[error]  * static method appendStmts(scala.collection.Seq,firrtl.ir.Statement)firrtl.ir.Statement in class firrtl.passes.memlib.VerilogMemDelays does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("firrtl.passes.memlib.VerilogMemDelays.appendStmts")
[error]  * static method replaceExp(scala.collection.mutable.HashMap,firrtl.ir.Expression)firrtl.ir.Expression in class firrtl.passes.memlib.VerilogMemDelays does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("firrtl.passes.memlib.VerilogMemDelays.replaceExp")
[error]  * static method memDelayStmt(scala.collection.mutable.HashMap,firrtl.Namespace,scala.collection.mutable.HashMap,scala.collection.mutable.ArrayBuffer,firrtl.ir.Statement)firrtl.ir.Statement in class firrtl.passes.memlib.VerilogMemDelays does not have a correspondent in current version                      
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("firrtl.passes.memlib.VerilogMemDelays.memDelayStmt")
[error]  * static method buildNetlist(scala.collection.mutable.HashMap,firrtl.ir.Statement)Unit in class firrtl.passes.memlib.VerilogMemDelays does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("firrtl.passes.memlib.VerilogMemDelays.buildNetlist")
[error]  * static method expToString(firrtl.ir.Expression)java.lang.String in class firrtl.passes.memlib.VerilogMemDelays does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("firrtl.passes.memlib.VerilogMemDelays.expToString")
[error]  * static method ug()firrtl.UnknownFlow# in class firrtl.passes.memlib.VerilogMemDelays does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("firrtl.passes.memlib.VerilogMemDelays.ug")
[error]  * object firrtl.passes.memlib.DelayPipe#PipeState does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[MissingClassProblem]("firrtl.passes.memlib.DelayPipe$PipeState$")
```

**Comment by jackkoenig** [comment]
We can filter them, just illustrating the kinds of issues we run in to. I'm going to PR the stuff that reports these errors.

**Comment by albert-magyar** [comment]
Yeah, this will require non-zero work to backport. Whenever we settle on a protocol I would be more than happy to open a second PR or just fix this one for the 1.2.1 branch.
